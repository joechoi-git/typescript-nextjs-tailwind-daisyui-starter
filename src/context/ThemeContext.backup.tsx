"use client";

import React, { createContext, useState, useContext, ReactNode, useEffect } from "react";

type Theme = "light" | "dark";

interface ThemeContextProps {
    theme: Theme;
    toggleTheme: () => void;
}

const ThemeContext = createContext<ThemeContextProps | undefined>(undefined);

export const useTheme = (): ThemeContextProps => {
    const context = useContext(ThemeContext);
    if (!context) {
        throw new Error("useTheme must be used within a ThemeProvider");
    }
    return context;
};

interface ThemeProviderProps {
    children: ReactNode;
}

export const ThemeProvider: React.FC<ThemeProviderProps> = ({ children }) => {
    const [theme, setTheme] = useState<Theme>("light");

    const toggleTheme = () => {
        setTheme((prevTheme) => (prevTheme === "light" ? "dark" : "light"));
    };

    useEffect(() => {
        document.body.className = theme;
    }, [theme]);

    return <ThemeContext.Provider value={{ theme, toggleTheme }}>{children}</ThemeContext.Provider>;
};
