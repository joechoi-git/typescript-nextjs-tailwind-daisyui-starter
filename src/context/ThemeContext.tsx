"use client";

import React from "react";

type Props = {
    children?: React.ReactNode;
};

type Theme =
    | "system"
    | "light"
    | "dark"
    | "cupcake"
    | "bumblebee"
    | "emerald"
    | "corporate"
    | "synthwave"
    | "retro"
    | "cyberpunk"
    | "valentine"
    | "halloween"
    | "garden"
    | "forest"
    | "aqua"
    | "lofi"
    | "pastel"
    | "fantasy"
    | "wireframe"
    | "black"
    | "luxury"
    | "dracula"
    | "cmyk"
    | "autumn"
    | "business"
    | "acid"
    | "lemonade"
    | "night"
    | "coffee"
    | "winter"
    | "dim"
    | "nord"
    | "sunset";

type IAuthContext = {
    theme: Theme;
    // eslint-disable-next-line no-unused-vars
    setTheme: (arg: Theme) => void;
};

const ThemeContext = React.createContext<IAuthContext>({
    theme: "system",
    setTheme: () => {}
});

const ThemeProvider = ({ children }: Props) => {
    // initialize context with cookie values
    const [theme, setTheme] = React.useState<Theme>("system");

    React.useEffect(() => {
        let savedTheme: Theme = localStorage.getItem("theme") as Theme;
        if (!savedTheme || savedTheme === "system") {
            if (window.matchMedia("(prefers-color-scheme: dark)").matches) {
                savedTheme = "dark";
            } else {
                savedTheme = "light";
            }
        }
        setTheme(savedTheme);
    }, []);

    return <ThemeContext.Provider value={{ theme, setTheme }}>{children}</ThemeContext.Provider>;
};

export { ThemeContext, ThemeProvider };
export type { Theme };
