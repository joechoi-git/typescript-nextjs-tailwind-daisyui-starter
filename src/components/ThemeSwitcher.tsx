import React from "react";
import { ThemeContext, Theme } from "../context/ThemeContext";

const ThemeSwitcher: React.FC = () => {
    const { theme, setTheme } = React.useContext(ThemeContext);
    const savedTheme: Theme | undefined =
        typeof window !== "undefined" ? (localStorage.getItem("theme") as Theme) : undefined;

    const handleSetTheme = (newTheme: Theme): void => {
        setTheme(newTheme);
        localStorage.setItem("theme", newTheme);
    };

    return (
        <select
            value={savedTheme ? savedTheme : theme}
            onChange={(e) => handleSetTheme(e.target.value as Theme)}
        >
            <option value="system">System</option>
            <option value="dark">Dark</option>
            <option value="light">Light</option>
            <option value="cupcake">Cupcake</option>
        </select>
    );
};

export default ThemeSwitcher;
