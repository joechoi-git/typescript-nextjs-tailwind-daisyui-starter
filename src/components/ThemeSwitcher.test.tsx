import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import ThemeSwitcher from "./ThemeSwitcher";
import { ThemeContext, Theme } from "../context/ThemeContext";

// Mock the ThemeContext
const mockSetTheme = jest.fn();

const renderWithContext = (theme: Theme) => {
    return render(
        <ThemeContext.Provider value={{ theme, setTheme: mockSetTheme }}>
            <ThemeSwitcher />
        </ThemeContext.Provider>
    );
};

describe("ThemeSwitcher", () => {
    beforeEach(() => {
        mockSetTheme.mockClear();
        localStorage.clear();
    });

    it("renders the theme switcher with the current theme selected", () => {
        renderWithContext("light");

        const select = screen.getByRole("combobox");
        expect(select).toBeInTheDocument();
        expect(select).toHaveValue("light");
    });

    it("changes the theme and updates localStorage when a new theme is selected", () => {
        renderWithContext("light");

        const select = screen.getByRole("combobox");
        fireEvent.change(select, { target: { value: "dark" } });

        expect(mockSetTheme).toHaveBeenCalledWith("dark");
        expect(localStorage.getItem("theme")).toBe("dark");
    });

    it("uses the saved theme from localStorage if available", () => {
        localStorage.setItem("theme", "dark");
        renderWithContext("light");

        const select = screen.getByRole("combobox");
        expect(select).toHaveValue("dark");
    });

    it("matches the snapshot", () => {
        const { asFragment } = renderWithContext("light");
        expect(asFragment()).toMatchSnapshot();
    });
});
