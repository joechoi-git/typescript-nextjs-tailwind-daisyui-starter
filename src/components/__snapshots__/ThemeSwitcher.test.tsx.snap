// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeSwitcher matches the snapshot 1`] = `
<DocumentFragment>
  <select
    class="border-solid border-2 p-2 border-primary"
  >
    <option
      value="system"
    >
      system
    </option>
    <option
      value="light"
    >
      light
    </option>
    <option
      value="dark"
    >
      dark
    </option>
    <option
      value="cupcake"
    >
      cupcake
    </option>
    <option
      value="bumblebee"
    >
      bumblebee
    </option>
    <option
      value="emerald"
    >
      emerald
    </option>
    <option
      value="corporate"
    >
      corporate
    </option>
    <option
      value="synthwave"
    >
      synthwave
    </option>
    <option
      value="retro"
    >
      retro
    </option>
    <option
      value="cyberpunk"
    >
      cyberpunk
    </option>
    <option
      value="valentine"
    >
      valentine
    </option>
    <option
      value="halloween"
    >
      halloween
    </option>
    <option
      value="garden"
    >
      garden
    </option>
    <option
      value="forest"
    >
      forest
    </option>
    <option
      value="aqua"
    >
      aqua
    </option>
    <option
      value="lofi"
    >
      lofi
    </option>
    <option
      value="pastel"
    >
      pastel
    </option>
    <option
      value="fantasy"
    >
      fantasy
    </option>
    <option
      value="wireframe"
    >
      wireframe
    </option>
    <option
      value="black"
    >
      black
    </option>
    <option
      value="luxury"
    >
      luxury
    </option>
    <option
      value="dracula"
    >
      dracula
    </option>
    <option
      value="cmyk"
    >
      cmyk
    </option>
    <option
      value="autumn"
    >
      autumn
    </option>
    <option
      value="business"
    >
      business
    </option>
    <option
      value="acid"
    >
      acid
    </option>
    <option
      value="lemonade"
    >
      lemonade
    </option>
    <option
      value="night"
    >
      night
    </option>
    <option
      value="coffee"
    >
      coffee
    </option>
    <option
      value="winter"
    >
      winter
    </option>
    <option
      value="dim"
    >
      dim
    </option>
    <option
      value="nord"
    >
      nord
    </option>
    <option
      value="sunset"
    >
      sunset
    </option>
  </select>
</DocumentFragment>
`;
